src/lib/error/bad-data-error.ts

```diff
@@ -1,23 +1,90 @@
-class BadDataError extends Error {
-    constructor(message: string) {
-        super();
-        Error.captureStackTrace(this, this.constructor);
+import { ErrorObject } from 'ajv';
+import { getPropFromString } from '../util/get-prop-from-string';
+import { ApiErrorSchema, UnleashError } from './api-error';
 
-        this.name = this.constructor.name;
-        this.message = message;
+type ValidationErrorDescription = {
+    description: string;
+    message: string;
+    path?: string;
+};
+class BadDataError extends UnleashError {
+    details: ValidationErrorDescription[];
+
+    constructor(
+        message: string,
+        errors?: [ValidationErrorDescription, ...ValidationErrorDescription[]],
+    ) {
+        const topLevelMessage =
+            'Request validation failed: your request body contains invalid data' +
+            (errors
+                ? '. Refer to the `details` list to see what happened.'
+                : `: ${message}`);
+        super(topLevelMessage);
+
+        this.details = errors ?? [{ message: message, description: message }];
     }
 
-    toJSON(): object {
+    toJSON(): ApiErrorSchema {
         return {
-            isJoi: true,
-            name: this.constructor.name,
-            details: [
-                {
-                    message: this.message,
-                },
-            ],
+            ...super.toJSON(),
+            details: this.details,
         };
     }
 }
 
 export default BadDataError;
+
+export const fromOpenApiValidationError =
+    (requestBody: object) =>
+    (validationError: ErrorObject): ValidationErrorDescription => {
+        // @ts-expect-error Unsure why, but the `dataPath` isn't listed on the type definition for error objects. However, it's always there. Suspect this is a bug in the library.
+        const propertyName = validationError.dataPath.substring(
+            '.body.'.length,
+        );
+        if (validationError.keyword === 'required') {
```

Comment:
I'd use "extract method" refactoring for each branch behind if/else if to explain the intent of each branch

Reply:
Done in b1b7a2938a22795dbf35aff385b07200b605f032